Makefile

```diff
@@ -78,9 +78,6 @@ endif
 
 all: debug
 
-install: release
```

Comment:
I think it's fine to keep a separate target.
Explicit `--offline --locked` is what we need to avoid it to be rebuilt (instead of relying on `$CARGO_INSTALL_OPTS`)

Reply:
Is there any advantage to have this two stage build & install for the `install` target? I'm feeling we would special case it and I'm not 100% sure `cargo install` is exactly (and will continue to be) equivalent to `cargo build --release && cargo install`.

Well, on the other hand when packaging it for a distro the two stage logic is used :)

Reply:
Follows the makefile logic. 
Also, `$CARGO_INSTALL_OPTS` was intentionally added/special-cased (part of the feedback) but now we're potentially mixing it with `$CARGO_CMD_OPTS`.
Considering the installation as a separate step (the options above allow to make it logically separate), I'm not sure it can be considered special casing. The only common thing is that `cargo` is used (which I guess is theoretically non-mandatory).
At some point install would need to install `-man`, `-completion` and so forth (of course, those will be different targets.
I'll leave it up to you, though.

Just to clarify, what I had in mind was something (unverified, simple enough to probably work):

```diff
diff --git a/Makefile b/Makefile
index 85feab7..72a37bc 100644
--- a/Makefile
+++ b/Makefile
@@ -1,11 +1,13 @@
 ROOT_DIR := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
-export LLC := llc
-export CLANG := clang
-export OBJCOPY := llvm-objcopy
+LLC := llc
+CLANG := clang
+OBJCOPY := llvm-objcopy
 
 CARGO := cargo
-RELEASE_VERSION ?= $(shell $(CARGO) metadata --no-deps --format-version=1 | jq -r '.packages | .[] | select(.name == "retis") | .version')
-RELEASE_NAME ?= $(shell $(CARGO) metadata --no-deps --format-version=1 | jq '.packages | .[] | select(.name=="retis") | .metadata.misc.release_name')
+RETIS_RELEASE_VERSION ?= $(shell $(CARGO) metadata --no-deps --format-version=1 | jq -r '.packages | .[] | select(.name == "retis") | .version')
+RETIS_RELEASE_NAME ?= $(shell $(CARGO) metadata --no-deps --format-version=1 | jq '.packages | .[] | select(.name=="retis") | .metadata.misc.release_name')
+
+export LLC CLANG OBJCOPY RETIS_RELEASE_NAME RETIS_RELEASE_VERSION
 
 PRINT = echo
 
@@ -79,13 +81,11 @@ endif
 all: debug
 
 install: release
-       $(CARGO) $(CARGO_OPTS) install $(CARGO_INSTALL_OPTS) --path=.
+       $(CARGO) $(CARGO_OPTS) install $(CARGO_INSTALL_OPTS) --path=. --offline --frozen
 
 define build
        $(call out_console,CARGO,$(strip $(2)) ...)
        $(Q)CARGO_BUILD_JOBS=$(CARGO_JOBS) \
-       RETIS_RELEASE_VERSION=$(RELEASE_VERSION) \
-       RETIS_RELEASE_NAME=$(RELEASE_NAME) \
        $(CARGO) $(CARGO_OPTS) $(1) $(CARGO_CMD_OPTS)
 endef
```

Also, we could consider, but it's really unrelated:

```
s/RETIS_RELEASE_NAME/RELEASE_NAME/
s/RETIS_RELEASE_VERSION/RELEASE_VERSION/
```

Reply:
> Follows the makefile logic. Also, `$CARGO_INSTALL_OPTS` was intentionally added/special-cased (part of the feedback) but now we're potentially mixing it with `$CARGO_CMD_OPTS`. 

Oops :(

> Considering the installation as a separate step (the options above allow to make it logically separate), I'm not sure it can be considered special casing. The only common thing is that `cargo` is used (which I guess is theoretically non-mandatory). At some point install would need to install `-man`, `-completion` and so forth (of course, those will be different targets.

Makes sense. We might even not use `cargo install` for some cases.

> Just to clarify, what I had in mind was something (unverified, simple enough to probably work):

Looks like there are many changes but I guess only the following is related:

```diff
 install: release
-       $(CARGO) $(CARGO_OPTS) install $(CARGO_INSTALL_OPTS) --path=.
+       $(CARGO) $(CARGO_OPTS) install $(CARGO_INSTALL_OPTS) --path=. --offline --frozen
```

Will do the above.

> Also, we could consider (unrelated):
> 
> ```
> s/RETIS_RELEASE_NAME/RELEASE_NAME/
> s/RETIS_RELEASE_VERSION/RELEASE_VERSION/
> ```

Right, it's not like we have different projects in the same repo :)

Reply:
Maybe we could default `CARGO_INSTALL_OPTS` to `--path=.` too?

Reply:
> > Follows the makefile logic. Also, `$CARGO_INSTALL_OPTS` was intentionally added/special-cased (part of the feedback) but now we're potentially mixing it with `$CARGO_CMD_OPTS`.
> 
> Oops :(
> 
> > Considering the installation as a separate step (the options above allow to make it logically separate), I'm not sure it can be considered special casing. The only common thing is that `cargo` is used (which I guess is theoretically non-mandatory). At some point install would need to install `-man`, `-completion` and so forth (of course, those will be different targets.
> 
> Makes sense. We might even not use `cargo install` for some cases.
> 
> > Just to clarify, what I had in mind was something (unverified, simple enough to probably work):
> 
> Looks like there are many changes but I guess only the following is related:
> 
> ```diff
>  install: release
> -       $(CARGO) $(CARGO_OPTS) install $(CARGO_INSTALL_OPTS) --path=.
> +       $(CARGO) $(CARGO_OPTS) install $(CARGO_INSTALL_OPTS) --path=. --offline --frozen
> ```
> 
> Will do the above.
> 

That's what I would have said, but seems that cargo is smart enough to detect that the env is different and then rebuilds accordingly. So it seems RELEASE_{VERSION,NAME} must be set.

> > Also, we could consider (unrelated):
> > ```
> > s/RETIS_RELEASE_NAME/RELEASE_NAME/
> > s/RETIS_RELEASE_VERSION/RELEASE_VERSION/
> > ```
> 
> Right, it's not like we have different projects in the same repo :)

Reply:
> That's what I would have said, but seems that cargo is smart enough to detect that the env is different and then rebuilds accordingly. So it seems RELEASE_{VERSION,NAME} must be set.

Ah, I see, missed that form the diff.

Reply:
> Maybe we could default `CARGO_INSTALL_OPTS` to `--path=.` too?

sorry, I missed this. Sounds good to me. If you decide to do so, would it be fine for you to add a note about this default value in the help target?